### PR TITLE
Removed _collapse_extra, simplified nested ifs

### PR DESCRIPTION
**Pull request by hanspi42**
#### References to other Issues or PRs
Part of #24561 

#### Brief description of what is fixed or changed
This does not change functionality, it just tidies the code up. The following has been done:

Several nested `if`s now use `or` and are easier to read.

Some to-do-related comments have been modified or removed.

Library code now has `simplify` and `dorational` as keyword-only arguments without defaults.

`_collapse_extra(self, extra)` has been removed, it is not used anymore in `laplace.py` and is not needed elsewhere.

`_laplace_deep_collect` has been re-written; a test case has been added.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by hanspi42** [comment]
@oscarbenjamin: One of the changes triggered an auto-request for you. Just tell us if you don't have the time.

File: sympy/integrals/laplace.py [hunk 1]
```diff
@@ -644,15 +645,14 @@ def _laplace_trig_expsum(f, t):
...
+        if (
+                (r := term.match(p*exp(m))) is not None
+                and (mp := r[m].as_poly(t)) is not None):
+            mc = mp.all_coeffs()
+            if len(mc) == 2:
+                xm.append({
+                    'k': r[p]*exp(mc[1]), 'a': mc[0],
+                    re: re(mc[0]), im: im(mc[0])})
             else:
                 xn.append(term)
```
Comment (oscarbenjamin) [inline sympy/integrals/laplace.py:657]: Neither of these is great but I think that the nested if is easier to read. Weird black-style formatting of complicated conditions is just a sign that something else needs to change.

I'd write this more like:
```python
if (r := term.match(p*exp(m))) is not None:
    if (mp := r[m].as_poly(t)) is not None:
        coeffs = mp.all_coeffs()
        if len(coeffs) == 2:
            c1, c2 = coeffs
            params = {'k': r[p]*exp(c2), 'a': c1, re: re(c1), im: im(c1)}
            xm.append(params)
        else:
            xn.append(term)
```
Then again I'm not sure why we need the nested matching and poly checking. Here we only accept a polynomial of two terms so can it not be part of the match?
```python
if (r := term.match(p*exp(c1*t + c2)) is not None):
```

**Comment by oscarbenjamin** [comment]
> Four private functions (_canonical, _canonical_coeff, _simplify, and _lin_eq2dict) that are used outside their modules now have names that do not start with `_`.

The main reason to prefix names with underscores is actually to signal to other codebases (outside of sympy) that these are not intended to be public and should not be used. It would be better if many more things had underscores for this reason.

Sometimes a function is defined in the wrong place but the fact that it is imported somewhere within the sympy codebase does not mean that it should have the underscore removed from its name.

**Comment by hanspi42** [comment]
> > Four private functions (_canonical, _canonical_coeff, _simplify, and _lin_eq2dict) that are used outside their modules now have names that do not start with `_`.
> 
> The main reason to prefix names with underscores is actually to signal to other codebases (outside of sympy) that these are not intended to be public and should not be used. It would be better if many more things had underscores for this reason.
> 
> Sometimes a function is defined in the wrong place but the fact that it is imported somewhere within the sympy codebase does not mean that it should have the underscore removed from its name.

Ah, then I went in the wrong direction. The much better way would then be to replicate those functions inside laplace.py. I'll do that.

File: sympy/integrals/laplace.py (hunk 1 above)
Comment (hanspi42) [reply sympy/integrals/laplace.py:657]: My problem is that this will not match

```
f = exp((t+a)*(t+b)-t**2)
```

I have to somehow deal with it.

**Comment by oscarbenjamin** [comment]
> The much better way would then be to replicate those functions inside laplace.py. I'll do that.

We don't want to duplicate the functions. There is nothing wrong with importing underscore prefixed function names within the sympy codebase.

File: sympy/integrals/laplace.py (hunk 1 above)
Comment (hanspi42) [reply sympy/integrals/laplace.py]: I now have a good solution for this, I think. I made a small modification to `_laplace_deep_collect`, and now it serves the purpose.

**Comment by hanspi42** [comment]
> We don't want to duplicate the functions. There is nothing wrong with importing underscore prefixed function names within the sympy codebase.

Thanks, now I have finally understood the underscore fully.

File: sympy/integrals/laplace.py (hunk 1 above)
Comment (hanspi42) [reply sympy/integrals/laplace.py]: One more change has happened, now it is completely ready for review.

**Comment by sylee957** [comment]
We could just use keyword-only arguments 

`def _laplace_transform(fn, t_, s_, *, simplify)` than keyword-arguments with defaults 
`def _laplace_transform(fn, t_, s_, simplify=False)`.

It is easy to make mistakes and defaults are redundant for the case too.

**Comment by hanspi42** [comment]
> We could just use keyword-only arguments

Very good idea.

**Comment by hanspi42** [comment]
> It is easy to make mistakes for accidentally using simplify=True default

There was actually such a mistake in the ILT ;)

**Pull request merged**